### ISSUE TITLE
Parsing MPLS error: "struct.error: unpack requires a buffer of 1 bytes"

### ISSUE THREAD

**Issue by lonecrane** [Guymer/PyGuymer3#4]
Hi, I have encountered a special blu-ray disc whose mpls is not parsed correctly by this code. However, the disc can be played by Potplayer and parsed by tsmuxer (one tool for demuxing) normally. Could you have a look at it? The relevant file is attached as follow
[00243.zip](https://github.com/Guymer/PyGuymer3/files/5334116/00243.zip)

**Comment by Guymer** [comment]
Thank for you noticing this and for providing a file that demonstrates the problem. I can replicate the behaviour on my local machine. My guess is that the function is reaching the end-of-file before it should (hence the zero/empty read). I will investigate further and see which length it is reading incorrectly. I will get back to you, thanks.

**Comment by Guymer** [comment]
It thinks that there are 51 audio streams but it reaches the end-of-file whilst reading the 40th one. Are you sure that this file has not become truncated somehow?

**Comment by Guymer** [comment]
It also appears that the file specification documentation that I used to make the functions (https://github.com/lw/BluRay/wiki/StreamEntry) has been updated since I wrote them, it is entirely possible that there is a parsing error in my package. I will go through it to see if anything has changed.

**Comment by Guymer** [comment]
Hi @lonecrane , I have spent a while today looking in to this for you and I think that PyGuymer3.MPLS is fine but that the MPLS file that you sent is not. Let me explain ...

I have updated every single function in PyGuymer3.MPLS after re-reading [the documentation](https://github.com/lw/BluRay/wiki/) - some more detail has been added since I created the functions in February 2018 (as part of the original Python 2.x module PyGuymer). Some of the updates were:
* The functions are now a lot more Pythonic and use `.seek()` and `.tell()` to jump around the MPLS file and read in all of the different parts.
* Some bugs were fixed and some new features were implemented in 02b14e3cec9c5cad590f440fb3dd1596764ced8b
* I have added warning messages for when values are not any of the expected ones, such as for the if-tests in [StreamAttributes](https://github.com/lw/BluRay/wiki/StreamAttributes).
* I have also added some debug prints (which are activated by passing `debug = True` to `pyguymer3.parse_MPLS_file()`) which print out where in the file the cursor currently is and how long each part of the file is.

These last 2 improvements have convinced me that the MPLS file that you sent is invalid and cannot be parsed by a standards compliant parser (as defined by [the documentation](https://github.com/lw/BluRay/wiki/)). If you now run `pyguymer3.parse_MPLS_file()` on the file you attached you will get this output:

```
WARNING: "StreamType" was not a recognised value 77
WARNING: "StreamCodingType" was not a recognised value 0
WARNING: "StreamType" was not a recognised value 192
WARNING: "StreamCodingType" was not a recognised value 51
WARNING: "StreamType" was not a recognised value 17
# and lots more lines
# ...
# ...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/foo/bar/pyguymer3/parse_MPLS_file.py", line 26, in parse_MPLS_file
    info["PlayList"] = load_PlayList(fobj, debug = debug)
  File "/foo/bar/pyguymer3/MPLS/load_PlayList.py", line 30, in load_PlayList
    ans["PlayItems"].append(load_PlayItem(fobj, debug = debug, indent = indent + 1))
  File "/foo/bar/pyguymer3/MPLS/load_PlayItem.py", line 47, in load_PlayItem
    ans["STNTable"] = load_STNTable(fobj, debug = debug, indent = indent + 1)
  File "/foo/bar/pyguymer3/MPLS/load_STNTable.py", line 40, in load_STNTable
    tmp["StreamAttributes"] = load_StreamAttributes(fobj, debug = debug, indent = indent + 1)
  File "/foo/bar/pyguymer3/MPLS/load_StreamAttributes.py", line 14, in load_StreamAttributes
    ans["Length"], = struct.unpack(">B", fobj.read(1))                          # [B]
struct.error: unpack requires a buffer of 1 bytes
```

You will see that there is an instance of "StreamType" within the MPLS file that has the value of 77 and that [the documentation for StreamType within StreamEntry](https://github.com/lw/BluRay/wiki/StreamEntry#streamtype) says that it can only be 1, 2, 3 or 4. Furthermore, if you run `pyguymer3.parse_MPLS_file()` with `debug = True` on the file you attached you will get this output:

```
DEBUG: pyguymer3.MPLS.load_header() called at 0 bytes and is 40 bytes long
DEBUG: pyguymer3.MPLS.load_AppInfoPlayList() called at 40 bytes and is 18 bytes long
DEBUG: pyguymer3.MPLS.load_PlayList() called at 58 bytes and is 1,444 bytes long
DEBUG:   pyguymer3.MPLS.load_PlayItem() called at 68 bytes and is 728 bytes long
DEBUG:     pyguymer3.MPLS.load_STNTable() called at 102 bytes and is 770 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 118 bytes and is 57 bytes long
# and lots more lines
# ...
# ...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/foo/bar/pyguymer3/parse_MPLS_file.py", line 26, in parse_MPLS_file
    info["PlayList"] = load_PlayList(fobj, debug = debug)
  File "/foo/bar/pyguymer3/MPLS/load_PlayList.py", line 30, in load_PlayList
    ans["PlayItems"].append(load_PlayItem(fobj, debug = debug, indent = indent + 1))
  File "/foo/bar/pyguymer3/MPLS/load_PlayItem.py", line 47, in load_PlayItem
    ans["STNTable"] = load_STNTable(fobj, debug = debug, indent = indent + 1)
  File "/foo/bar/pyguymer3/MPLS/load_STNTable.py", line 40, in load_STNTable
    tmp["StreamAttributes"] = load_StreamAttributes(fobj, debug = debug, indent = indent + 1)
  File "/foo/bar/pyguymer3/MPLS/load_StreamAttributes.py", line 14, in load_StreamAttributes
    ans["Length"], = struct.unpack(">B", fobj.read(1))                          # [B]
struct.error: unpack requires a buffer of 1 bytes
```

Note the two lines `pyguymer3.MPLS.load_PlayItem() called at 68 bytes and is 728 bytes long` and `pyguymer3.MPLS.load_STNTable() called at 102 bytes and is 770 bytes long` - this is the problem. Not only does the MPLS file that you attached contain undocumented values for "StreamType" but the "STNTable" is larger than the "PlayItem" that contains it (770 is greater than 728). This is the root cause of the problem here: the file is too short, there are child blocks within the file that are larger than the parents that contain them.

If I run `pyguymer3.parse_MPLS_file()` with `debug = True` on a different MPLS file that I have then this is the output that I get:

```
DEBUG: pyguymer3.MPLS.load_header() called at 0 bytes and is 40 bytes long
DEBUG: pyguymer3.MPLS.load_AppInfoPlayList() called at 40 bytes and is 18 bytes long
DEBUG: pyguymer3.MPLS.load_PlayList() called at 58 bytes and is 780 bytes long
DEBUG:   pyguymer3.MPLS.load_PlayItem() called at 68 bytes and is 730 bytes long
DEBUG:     pyguymer3.MPLS.load_STNTable() called at 102 bytes and is 696 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 118 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 128 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 134 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 144 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 150 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 160 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 166 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 176 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 182 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 192 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 198 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 208 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 214 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 224 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 230 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 240 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 246 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 256 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 262 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 272 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 278 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 288 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 294 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 304 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 310 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 320 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 326 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 336 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 342 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 352 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 358 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 368 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 374 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 384 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 390 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 400 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 406 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 416 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 422 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 432 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 438 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 448 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 454 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 464 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 470 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 480 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 486 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 496 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 502 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 512 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 518 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 528 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 534 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 544 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 550 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 560 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 566 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 576 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 582 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 592 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 598 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 608 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 614 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 624 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 630 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 640 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 646 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 656 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 662 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 672 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 678 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 688 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 694 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 704 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 710 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 720 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 726 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 736 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 742 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 752 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 758 bytes and is 1 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 759 bytes and is 1 bytes long
DEBUG:   pyguymer3.MPLS.load_SubPath() called at 798 bytes and is 40 bytes long
DEBUG:     pyguymer3.MPLS.load_SubPlayItem() called at 808 bytes and is 30 bytes long
DEBUG: pyguymer3.MPLS.load_PlayListMark() called at 838 bytes and is 1,126 bytes long
DEBUG: pyguymer3.MPLS.load_ExtensionData() called at 1,964 bytes and is 1,870 bytes long
```

Note how all of the children are shorter than the parents, also, the total read bytes is equal to the file size.

In short, I think that the MPLS file you have sent me is broken. The rest of the Blu-ray that the MPLS file came from might be absolutely fine and it might be that just this one MPLS file became corrupted. I cannot see a problem with what PyGuymer3.MPLS is doing here, I think that it is following [the documentation](https://github.com/lw/BluRay/wiki/) correctly. Do you (dis)agree?

**Comment by lonecrane** [comment]
@Guymer Thanks a lot for your hard work. This mpls file should be correct, as it can be used for playback with potplayer and for demuxing with tsmuxer.
![Untitled2](https://user-images.githubusercontent.com/36885830/95659041-20792980-0b51-11eb-9f64-a54b8cccbfe2.png)
I opened this file by a hex editor. As shown, this mpls seems to have extra data "00337M2TS.00338M2TS", following the first playitem "00001M2TS". The 2nd playitem (00336M2TS) starts at 0x310. It's a pity that I know little about multimedia, and is the abnormal data called "multiangle"?

**Comment by lonecrane** [comment]
@Guymer Dear Guymer, it seems that the bug lies in parsing the value of 'IsMultiAngle'. The provided mpls indeed contains multi angles. Comparing to the mpls-parsing code of tsmuxer [https://github.com/justdan96/tsMuxer/blob/master/tsMuxer/tsPacket.cpp, MPLSParser::parsePlayItem(), Line 2160], the current python code 
`ans["IsMultiAngle"] = bool(ans["MiscFlags1"]&(1<<(11)))` 
should be 
`ans["IsMultiAngle"] = bool(ans["MiscFlags1"]&(1<<(16-1-11)))` 
to skip the first 11 bits correctly. With this correction and removing the subsequent 'decode('utf-8')' in 
```
                tmp["ClipInformationFileName"] = fobj.read(5).decode('utf-8')
                tmp["ClipCodecIdentifier"] = fobj.read(4).decode('utf-8')
```
I have got the right results.

**Comment by Guymer** [comment]
Hi @lonecrane ,

Yes, you are quite right - that is a bug. I was not correctly taking account of the endianness of the integer when I was extracting the "IsMultiAngle" boolean value, thank you very much for spotting that, it has now been fixed in f748764b60b0ed70b0382e76d4ecc1bb1e71bc81

As for the second issue of it now failing due to the `.decode()`, that is less clear in my opinion. I agree that the MPLS file is now parsed correctly except for one of those `.decode()` calls, but I am not sure that is a bug in PyGuymer3.MPLS but rather, once again, a problem with the file. [The documentation for ClipInformationFileName within PlayItem](https://github.com/lw/BluRay/wiki/PlayItem#clipinformationfilename-clipcodecidentifier-and-reftostcid) says that:

```
The first two fields contain an ASCII-compatible string of, respectively, 5 and 4 characters. The first one should contain only values '0'..'9' and the second should contain the value "M2TS".
```

If I put a `print(tmp)` statement in that bit of code in "pyguymer3/MPLS/load_PlayItem.py" (after removing the `.decode()` bits) then I see that there are 3 angles with the following information:

```
{'ClipInformationFileName': b'00337', 'ClipCodecIdentifier': b'M2TS', 'RefToSTCID': 0}
{'ClipInformationFileName': b'00338', 'ClipCodecIdentifier': b'M2TS', 'RefToSTCID': 0}
{'ClipInformationFileName': b'\x02\x9e\x00\x00\x01', 'ClipCodecIdentifier': b'\t\x1f\x00\x00', 'RefToSTCID': 0}
```

The third angle's "ClipInformationFileName" is not '0'..'9' and the third's "ClipCodecIdentifier" is not "M2TS". Is this what you expect that third angle to look like? Is there truly a clip file called "\x02\x9e\x00\x00\x01.mpls" on the Blu-ray disc?

In the meantime, to be more tolerant of decoding errors, I have now implemented a "errors" keyword argument to all of the functions in e90877b2f6db539d8faaeb9cc2bf2d9ad17556c7. This is simply a pass-through for the "errors" keyword argument of `bytes.decode`, as explained in [the documentation](https://docs.python.org/3/library/stdtypes.html?highlight=decode#bytes.decode). Therefore, if you now call `pyguymer3.parse_MPLS_file()` with `debug = True` and `errors = "replace"` then it will run successfully and you should get this output:

```
DEBUG: pyguymer3.MPLS.load_header() called at 0 bytes and is 40 bytes long
DEBUG: pyguymer3.MPLS.load_AppInfoPlayList() called at 40 bytes and is 18 bytes long
DEBUG: pyguymer3.MPLS.load_PlayList() called at 58 bytes and is 1,444 bytes long
DEBUG:   pyguymer3.MPLS.load_PlayItem() called at 68 bytes and is 728 bytes long
DEBUG:     pyguymer3.MPLS.load_STNTable() called at 134 bytes and is 2 bytes long
DEBUG:   pyguymer3.MPLS.load_PlayItem() called at 796 bytes and is 706 bytes long
DEBUG:     pyguymer3.MPLS.load_STNTable() called at 830 bytes and is 672 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 846 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 856 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 862 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 872 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 878 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 888 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 894 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 904 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 910 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 920 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 926 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 936 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 942 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 952 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 958 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 968 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 974 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 984 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 990 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,000 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,006 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,016 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,022 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,032 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,038 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,048 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,054 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,064 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,070 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,080 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,086 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,096 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,102 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,112 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,118 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,128 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,134 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,144 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,150 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,160 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,166 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,176 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,182 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,192 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,198 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,208 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,214 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,224 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,230 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,240 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,246 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,256 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,262 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,272 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,278 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,288 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,294 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,304 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,310 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,320 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,326 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,336 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,342 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,352 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,358 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,368 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,374 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,384 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,390 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,400 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,406 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,416 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,422 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,432 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,438 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,448 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,454 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,464 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,470 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,480 bytes and is 6 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamEntry() called at 1,486 bytes and is 10 bytes long
DEBUG:       pyguymer3.MPLS.load_StreamAttributes() called at 1,496 bytes and is 6 bytes long
DEBUG: pyguymer3.MPLS.load_PlayListMark() called at 1,502 bytes and is 230 bytes long
DEBUG: pyguymer3.MPLS.load_ExtensionData() called at 1,732 bytes and is 88 bytes long
```

Given that the file is 1,820 bytes (= 1,732 + 88) long then I think that the "IsMultiAngle" bug fix has made it work now, thank you for that.

**Comment by lonecrane** [comment]
@Guymer Thanks a lot. By comparing to the code of tsmuxer again, I found that the second bug lies in the start index of the multi angle counting. The first angle, whose index is 0, has always been parsed previously, so the subsequent parsing should start from 1. And the decode() method is necessary to get the right results.

**Comment by Guymer** [comment]
You are correct, thank you very much for spotting that. I have now fixed that in d8a4c416f529ee647f302616bfe2a79d6b334f1f

Thanks for your help @lonecrane

**Comment by Guymer** [comment]
(I have just updated [the documentation for PlayItem](https://github.com/lw/BluRay/wiki/PlayItem) too)
